hugr-persistent/src/wire.rs

```diff
@@ -0,0 +1,287 @@
+        PersistentWire::from_port(node, port, self)
+    }
+
+    /// Whether a node is valid in `self`, is deleted or is an IO node in a
+    /// replacement graph.
+    fn node_status(&self, PatchNode(commit_id, node): PatchNode) -> NodeStatus {
+        debug_assert!(self.contains_id(commit_id), "unknown commit");
+        if self
+            .replacement(commit_id)
+            .is_some_and(|repl| repl.get_replacement_io().contains(&node))
+        {
+            NodeStatus::InOut
+        } else if self.deleted_nodes(commit_id).contains(&node) {
+            NodeStatus::Deleted
+        } else {
+            NodeStatus::Valid
+        }
+    }
+}
+
+impl PersistentWire {
+    /// Get the wire connected to a specified port of a pinned node in `hugr`.
+    pub fn from_port<R>(
+        node: PatchNode,
+        port: impl Into<Port>,
+        per_hugr: &PersistentHugr<R>,
+    ) -> Self {
+        assert!(per_hugr.contains_node(node), "node not in hugr");
+
+        let mut wires = BTreeSet::from_iter([to_wire(node, port, per_hugr)]);
+        let mut queue = VecDeque::from_iter(wires.iter().copied());
+
+        while let Some(wire) = queue.pop_front() {
+            let all_ports = all_ports(wire, per_hugr);
+
+            for (per_node @ PatchNode(commit_id, node), port) in all_ports {
+                let hugr = per_hugr.commit_hugr(commit_id);
+                match per_hugr.node_status(per_node) {
+                    NodeStatus::Deleted => {
+                        // If node is deleted, check if there are wires between
+                        // ports on the opposite end of the wire and boundary
+                        // ports in the child commit that deleted the node.
+                        let deleted_by = per_hugr
+                            .find_deleting_commit(per_node)
+                            .expect("deleted node has deleting commit");
+                        for (opp_node, opp_port) in hugr.linked_ports(node, port) {
+                            let opp_node = PatchNode(commit_id, opp_node);
+                            for (child_node, child_port) in per_hugr
+                                .as_state_space()
+                                .linked_child_ports(opp_node, opp_port, deleted_by)
+                            {
+                                let w = to_wire(child_node, child_port, per_hugr);
+                                if wires.insert(w) {
+                                    queue.push_back(w);
+                                }
+                            }
+                        }
+                    }
+                    NodeStatus::InOut => {
+                        // If node is an IO node in a replacement graph, there
+                        // must be (at least) one wire
+                        // between the boundary ports of the
+                        // commit (the ports opposite of the IO node) and ports
```

Comment:
```suggestion
                        // commit and ports
```

"Opposite" meaning that if the node is an Input, there are wires from the output-boundary (to ports in a parent commit)...right? I kinda get what you mean by "opposite" ("furthest away" in a nested-DFG picture with wires flowing from top to bottom kind of thing), but I think it'd be good to rule out the other interpretation (opposite direction) more strongly